config/file.go

```diff
@@ -0,0 +1,323 @@
+	fs.invokeConfigListeners(oldCfg, newCfg)
+
+	return oldCfg, nil
+}
+
+// Patch merges the given config with the current configuration, without updating the backing store.
+func (fs *fileStore) Patch(*model.Config) (*model.Config, error) {
+	fs.configLock.Lock()
+	defer fs.configLock.Unlock()
+
+	panic("patch is not yet implemented")
+
+	return fs.config, nil
+}
+
+// serialize converts the given configuration into JSON bytes for persistence.
+func (fs *fileStore) serialize(cfg *model.Config) ([]byte, error) {
+	return json.MarshalIndent(cfg, "", "    ")
+}
+
+// persist writes the configuration to the configured file.
+func (fs *fileStore) persist(cfg *model.Config) error {
+	fs.stopWatcher()
+
+	b, err := fs.serialize(cfg)
+	if err != nil {
+		return errors.Wrap(err, "failed to serialize")
+	}
+
+	err = ioutil.WriteFile(fs.path, b, 0644)
+	if err != nil {
+		return errors.Wrap(err, "failed to write file")
+	}
+
+	fs.needsSave = false
+	if fs.watch {
+		if err = fs.startWatcher(); err != nil {
+			mlog.Error("failed to start config watcher", mlog.String("path", fs.path), mlog.Err(err))
+		}
+	}
+
+	return nil
+}
+
+// Load updates the current configuration from the backing store.
+func (fs *fileStore) Load() (needsSave bool, err error) {
+	var f io.ReadCloser
+
+	f, err = os.Open(fs.path)
+	if os.IsNotExist(err) {
+		needsSave = true
+		defaultCfg := model.Config{}
+		defaultCfg.SetDefaults()
+
+		var defaultCfgBytes []byte
+		defaultCfgBytes, err = fs.serialize(&defaultCfg)
+		if err != nil {
+			return false, errors.Wrap(err, "failed to serialize default config")
+		}
+
+		f = ioutil.NopCloser(bytes.NewReader(defaultCfgBytes))
+
+	} else if err != nil {
+		return false, errors.Wrapf(err, "failed to open %s for reading", fs.path)
+	}
+	defer func() {
+		closeErr := f.Close()
+		if err == nil && closeErr != nil {
+			err = errors.Wrap(closeErr, "failed to close")
+		}
+	}()
+
+	allowEnvironmentOverrides := true
+	loadedCfg, environmentOverrides, err := readConfig(f, allowEnvironmentOverrides)
+	if err != nil {
+		return false, errors.Wrapf(err, "failed to load config from %s", fs.path)
+	}
+
+	// SetDefaults generates various keys and salts if not previously configured. Determine if
+	// such a change will be made before invoking. This method will not effect the save: that
+	// remains the responsibility of the caller.
+	needsSave = needsSave || loadedCfg.SqlSettings.AtRestEncryptKey == nil || len(*loadedCfg.SqlSettings.AtRestEncryptKey) == 0
+	needsSave = needsSave || loadedCfg.FileSettings.PublicLinkSalt == nil || len(*loadedCfg.FileSettings.PublicLinkSalt) == 0
+	needsSave = needsSave || loadedCfg.EmailSettings.InviteSalt == nil || len(*loadedCfg.EmailSettings.InviteSalt) == 0
+
+	loadedCfg.SetDefaults()
+
+	if err := loadedCfg.IsValid(); err != nil {
+		return false, errors.Wrap(err, "invalid config")
+	}
+
+	if changed := fixConfig(loadedCfg); changed {
+		needsSave = true
+	}
+
+	fs.configLock.Lock()
```

Comment:
Note that we haven't had to acquire a lock until now in this method, because we haven't read or written to anything mutable on the file store. `path` doesn't count, since it can't be modified after initialization.